numpy/core/src/umath/ufunc_object.c

```diff
@@ -2739,8 +2740,41 @@ reducelike_promote_and_resolve(PyUFuncObject *ufunc,
         PyArrayObject *arr, PyArrayObject *out,
         PyArray_DTypeMeta *signature[3],
         npy_bool enforce_uniform_args, PyArray_Descr *out_descrs[3],
-        char *method)
+        NPY_CASTING casting, char *method)
 {
+     /*
+      * If no dtype is specified and out is not specified, we override the
+      * integer and bool dtype used for add and multiply.
+      *
+      * TODO: The following should be handled by a promoter!
+      */
+    if (signature[0] == NULL && out == NULL) {
+        /*
+         * For integer types --- make sure at least a long
```

Comment:
`long` changes size depending on the platform, would it not be better to be explicit?

Reply:
Yes, but this code is just moved.  I.e. changing it is probably tied to thinking about changing the default integer type (although this path for `sum` and `prod` could be changed separately maybe)/